File: day02/keypad2.pl
#!/usr/bin/env perl
use strict;
use warnings;
use List::Util qw(min max);
use 5.24.0;

my $keypad = [
              [undef, undef, undef, undef, undef, undef, undef],
              [undef, undef, undef,     1, undef, undef, undef],
              [undef, undef,     2,     3,     4, undef, undef],
              [undef,     5,     6,     7,     8,     9, undef],
              [undef, undef,   'A',   'B',   'C', undef, undef],
              [undef, undef, undef,   'D', undef, undef, undef],
              [undef, undef, undef, undef, undef, undef, undef],
             ];

my ($y, $x) = (3, 1);
my @code;

while (<>) {
    chomp;
    my @moves = split //;
    for my $move (@moves) {
        if ($move eq 'U' && defined $keypad->[$y-1][$x]) {
            $y--;
        }
        if ($move eq 'D' && defined $keypad->[$y+1][$x]) {
            $y++;
        }
        if ($move eq 'L' && defined $keypad->[$y][$x-1]) {
            $x--;
        }
        if ($move eq 'R' && defined $keypad->[$y][$x+1]) {
            $x++;
        }
#        say "$move $x $y $keypad->[$y][$x]";
    }
#    say "adding $keypad->[$y][$x]";
    push @code, $keypad->[$y][$x];
}

say join "", @code;
# say scalar @code;
# for my $i (0..$#code) {
#     say "$i $code[$i]";
# }

File: day02/keypad2.py
#!/usr/bin/env python
from sys import argv

keypad = [
    [0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 1, 0, 0, 0],
    [0, 0, 2, 3, 4, 0, 0],
    [0, 5, 6, 7, 8, 9, 0],
    [0, 0,'A','B','C', 0, 0],
    [0, 0, 0, 'D', 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0] ]

(x, y) = (3, 1)
code = []

script, filename = argv
with open(filename) as f :
    for line in f :
        line = line.rstrip("\n")
        for move in line :
            if move == 'U' and keypad[y-1][x] != 0 :
                y = y - 1
            elif move == 'D' and keypad[y+1][x] != 0 :
                y = y + 1
            elif move == 'L' and keypad[y][x-1] != 0 :
                x = x - 1
            elif move == 'R' and keypad[y][x+1] != 0:
                x = x + 1

        code.append(keypad[y][x])

print ''.join(map(str,code))
